design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,345 @@
+indicate an error during the translation of the call and not to indicate errors from components. Components are
+encouraged to avoid returning errors as discussed later in [Alternative to errors](#alternative-to-errors).
+
+If the error being mapped has an equivalent in the set of errors and the projection of the consumer's
+language provides an equivalent mapping from that mapped error, then that error will have a natural
+mapping in the other language. In the case there isn't an equivalent error to map to in xlang, then a
+generic error, xlang_fail, will be available to be used. Note that in this case, the natural mapping of
+errors from one language to the other will be lost in the sense that it will not be an equivalent mapping.
+In the case there isn't an equivalent error to map to in a consumer's language, it is left to the projection
+of that language to decide what error from its language to map to based on what makes sense for that language.
+This can end up being a generic error.
+
+## xlang errors
+
+xlang will use an Int32, xlang_result, to represent its errors. The representation of this 32-bit
+value will be based on the same model used in COM for HRESULTs to assist with sharing error codes,
+but will not necessarily be defined in this design note on a bit by bit basis as is in COM.
+This is because xlang will not inherit all the errors used in COM and will instead define a set of
+errors with their values defined. This will allow xlang to avoid defining all the errors available
+in COM on each supported platform especially when most will not be used and to scope its errors to ones
+that are likely to occur during translation of calls. This will also allow for projections to avoid
+needing to maintain a large mapping of errors.
+
+The following errors will be defined in xlang:
+
+xlang error                | Value      | Description
+-------------------------- | ---------- | ----------------------------------------------------------------------
+xlang_access_denied        | 0x80070005 | Requested operation / call is not allowed
+xlang_bounds               | 0x8000000b | Invalid index used such as when indexing arrays
+xlang_fail                 | 0x80004005 | Generic error when no equivalent or better error mapping is available
+xlang_handle               | 0x80000006 | Invalid handle
+xlang_invalid_arg          | 0x80000003 | Invalid argument in a call
+xlang_no_interface         | 0x80000004 | Class or interface not found
+xlang_not_impl             | 0x80000001 | Function / feature is not implemented
+xlang_out_of_memory        | 0x80000002 | Allocation issue resulting from not enough memory
+xlang_pointer              | 0x80000005 | Null pointer related errors
+xlang_type_load            | 0x80131522 | Syntax / type error during runtime
+
+Even though xlang only defines these errors, it will not explicitly prevent the use of other errors
```

Comment:
As per arch meeting, reserve 0 as success. 

Also, we had an interesting discussion about the need for minting further error codes. Generally we have a philosophy of not wanting to use exceptions for flow control. There are suggestions for separating what we called "domain errors" from more general runtime/language errors.

Reply:
Addressed.